Add TTIR verification

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119599
* #119859
* #119838
* #119836
* __->__ #119835
* #119834

Make sure the TTIR generated is valid before attempting to analyze. Incorrectly written triton code would produce broken TTIR. Minor discussion on https://github.com/openai/triton/issues/3120

oulgen committed:
Add TTIR verification

[ghstack-poisoned]

oulgen committed:
Update on "Add TTIR verification"


Make sure the TTIR generated is valid before attempting to analyze. Incorrectly written triton code would produce broken TTIR. Minor discussion on https://github.com/openai/triton/issues/3120

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -154,6 +154,8 @@ def generate_ttir(kernel, kwargs):
 
     src = ASTSource(kernel, signature, constants, specialization)
     ttir_module = src.make_ir(options, context)
+    if not ttir_module.verify():
+        raise Exception("Verification for TTIR module has failed")
```
What is this? We can get malformed TTIR from Triton?

oulgen replied:
If the user code is malformed, then triton produced malformed TTIR which gets checked just like this at a later time

oulgen committed:
Update on "Add TTIR verification"


Make sure the TTIR generated is valid before attempting to analyze. Incorrectly written triton code would produce broken TTIR. Minor discussion on https://github.com/openai/triton/issues/3120

[ghstack-poisoned]
